### PR TITLE
Drop Java 7 support and add try methods

### PR DESCRIPTION
**Pull request by oschwald**
This includes #147.

### CODE REVIEW

File: src/main/java/com/maxmind/geoip2/model/AbstractCityResponse.java [hunk 1]
```diff
@@ -33,7 +33,7 @@
         this.city = city != null ? city : new City();
         this.location = location != null ? location : new Location();
         this.postal = postal != null ? postal : new Postal();
-        this.subdivisions = subdivisions != null ? subdivisions : new ArrayList<Subdivision>();
+        this.subdivisions = subdivisions != null ? subdivisions : new ArrayList<>();
     }
```
[the remark this answers is not in the document]
Comment (oschwald) [reply src/main/java/com/maxmind/geoip2/model/AbstractCityResponse.java]: These were removed as their are noise and are no longer needed as of Java 7.

File: src/main/java/com/maxmind/geoip2/DatabaseReader.java [hunk 2]
```diff
@@ -167,12 +174,34 @@ public DatabaseReader build() throws IOException {
...
+     * @return A <T> object with the data for the IP address or null if the IP address is not in our database
+     * @throws IOException              if there is an error opening or reading from the file.
+     */
+    private <T> Optional<T> get(InetAddress ipAddress, Class<T> cls,
+                      String type, int stackDepth) throws IOException, AddressNotFoundException {
 
         String databaseType = this.getMetadata().getDatabaseType();
         if (!databaseType.contains(type)) {
-            String caller = Thread.currentThread().getStackTrace()[2]
+            String caller = Thread.currentThread().getStackTrace()[2 + stackDepth]
```
[the remark this answers is not in the document]
Comment (oschwald) [reply src/main/java/com/maxmind/geoip2/DatabaseReader.java]: Could you rephrase the question? This is necessary as the depth now varies depending on whether it is called from the try methods or `getOrThrowException`.

[the remark this answers is not in the document]
Comment (eilara) [reply src/main/java/com/maxmind/geoip2/DatabaseReader.java]: Gotcha. So this required no changes to existing tests, if it is even tested.

File: src/main/java/com/maxmind/geoip2/model/AbstractCityResponse.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (eilara) [reply src/main/java/com/maxmind/geoip2/model/AbstractCityResponse.java]: Right, misread that. So this is for `In Java SE 7 and later, you can replace the type arguments required to invoke the constructor of a generic class with an empty set of type arguments (<>) as long as the compiler can ...` (from jdk docs)

File: src/main/java/com/maxmind/geoip2/DatabaseReader.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (oschwald) [reply src/main/java/com/maxmind/geoip2/DatabaseReader.java]: It is tested at https://github.com/maxmind/GeoIP2-java/blob/6477ec2724d5f2c338554fc8a1befa4063fc60b3/src/test/java/com/maxmind/geoip2/DatabaseReaderTest.java#L183

**Comment by maxgv** [comment]
thank


| |
马新港
|
|
ma342221@163.com
|
签名由网易邮箱大师定制
On 10/17/2019 05:05，Ran Eilam<notifications@github.com> wrote：

@eilara commented on this pull request.

In src/main/java/com/maxmind/geoip2/model/AbstractCityResponse.java:

> @@ -33,7 +33,7 @@
         this.city = city != null ? city : new City();
         this.location = location != null ? location : new Location();
         this.postal = postal != null ? postal : new Postal();
-        this.subdivisions = subdivisions != null ? subdivisions : new ArrayList<Subdivision>();
+        this.subdivisions = subdivisions != null ? subdivisions : new ArrayList<>();
     }


Right, misread that. So this is for In Java SE 7 and later, you can replace the type arguments required to invoke the constructor of a generic class with an empty set of type arguments (<>) as long as the compiler can ... (from jdk docs)
